Index: defaults.make
===================================================================
--- defaults.make	(revision 7227)
+++ defaults.make	(working copy)
@@ -35,7 +35,7 @@
 
 # basic parameters
 
-BINARYVERSION ?= 3
+BINARYVERSION = 3
 NURSERY ?= (128*1024)
 STACKDIRECTION ?= 1
 CROSS_CHICKEN ?= 0
@@ -221,10 +221,10 @@
 
 # file extensions
 
-O ?= .o
-A ?= .a
+O = .o
+A = .a
 # EXE =
-SO ?= .so
+SO = .so
 
 # special files
 
Index: chicken-setup.scm
===================================================================
--- chicken-setup.scm	(revision 7227)
+++ chicken-setup.scm	(working copy)
@@ -146,6 +146,9 @@
 (define *target-libs* (foreign-value "C_TARGET_MORE_LIBS" c-string))
 (define *target-lib-home* (foreign-value "C_TARGET_LIB_HOME" c-string))
 
+(define *major-version* (##sys#fudge 41))
+(define *default-eggdir* (conc "eggs/" *major-version*))
+
 (define *windows*
   (and (eq? (software-type) 'windows) 
        (build-platform) ) )
@@ -187,7 +190,8 @@
 (define *svn-repository* #f)
 (define *local-repository* #f)
 (define *destdir* #f)
-(define *repository-hosts* '(("www.call-with-current-continuation.org" "eggs" 80)))
+(define *repository-hosts*
+  (list (list "www.call-with-current-continuation.org" *default-eggdir* 80)))
 (define *revision* #f)
 (define *run-tests* #f)
 (define *fetched-eggs* '())
@@ -1140,15 +1144,17 @@
 			  )))
     (with-output-to-file (doc-index)
       (lambda ()
-	(printf "<html><head><title>Egg documentation index for ~a</title><link rel=\"stylesheet\" type=\"text/css\" href=\"style.css\"/></head>~%" hn)
-	(printf "<body><a id=\"official-index\" href=\"http://www.call-with-current-continuation.org/eggs/index.html\">Visit the official egg index</a>~%")
-	(printf "<h1 id=\"title\">Egg documentation index:</h1>~%")
+	(print "<html><head><title>Egg documentation index for " hn 
+	       "</title><link rel=\"stylesheet\" type=\"text/css\" href=\"style.css\"/></head>")
+	(print "<body><a id=\"official-index\" href=\"http://www.call-with-current-continuation.org/"
+	       *default-eggdir* "/index.html\">Visit the official egg index</a>")
+	(print "<h1 id=\"title\">Egg documentation index:</h1>")
 	(printf "<p id=\"install-info\">CHICKEN: ~a<br>Host: ~a<br>Repository path: ~a<br><p>~%" 
 		(chicken-version #t)
 		(get-host-name)
 		rpath)
-	(printf "<table id=\"egg-index\">~%")
-	(printf "<thead><tr><th>Egg name</th><th>Version</th><th>Release</th></tr></thead>~%<tbody>~%")
+	(print "<table id=\"egg-index\">")
+	(print "<thead><tr><th>Egg name</th><th>Version</th><th>Release</th></tr></thead>\n<tbody>")
 	(let ((c 0))
 	  (for-each
 	   (lambda (f)
@@ -1238,8 +1244,8 @@
   (define (parse-host host eggdir)
     (set! *repository-hosts*
       (cons (match (string-match "(.+)\\:([0-9]+)" host)
-	      ((_ host port) (list host (if eggdir "eggs" "") (string->number port)))
-	      (_ (list host (if eggdir "eggs" "") 80)) )
+	      ((_ host port) (list host (if eggdir *default-eggdir* "") (string->number port)))
+	      (_ (list host (if eggdir (conc *default-eggdir* "") 80)) ) )
 	    *repository-hosts*) )  )
   (setup-root-directory *base-directory*)
   (let ((uinst #f)
Index: chicken.h
===================================================================
--- chicken.h	(revision 7227)
+++ chicken.h	(working copy)
@@ -39,6 +39,8 @@
 #ifndef ___CHICKEN
 #define ___CHICKEN
 
+#define C_MAJOR_VERSION       3
+
 /*
  * N.B. This file MUST not rely upon "chicken-config.h"
  */
Index: misc/setversion.scm
===================================================================
--- misc/setversion.scm	(revision 7227)
+++ misc/setversion.scm	(working copy)
@@ -1,6 +1,6 @@
 ;;;; setversion.scm - Bump version-number
 
-(use utils)
+(use srfi-1 utils)
 
 (define buildversion (->string (car (read-file "buildversion"))))
 (define buildbinaryversion (car (read-file "buildbinaryversion")))
@@ -25,11 +25,16 @@
        (patch (list both tmp) rx subst)
        (system* "mv ~S ~S" tmp both ) ) ) ) )
 
+(define (parse-version v)
+  (string-match "(\\d+)\\.(\\d+)\\.(\\d+)(.*)" v) )
+
 (define (main args)
   (cond ((member "-set" args) =>
 	 (lambda (a) (set! buildversion (cadr a))) )
 	((not (member "-noinc" args))
-	 (set! buildversion (number->string (+ (string->number buildversion) 0.001))) ) )
+	 (match (parse-version buildversion)
+	   ((_ maj min pl huh) 
+	    (set! buildversion (conc maj "." min "." (add1 (string->number pl)) huh)) ) ) ) )
   (with-output-to-file "buildversion" (cut display buildversion))
   (with-output-to-file "version.scm" 
     (lambda ()
Index: version.scm
===================================================================
--- version.scm	(revision 7227)
+++ version.scm	(working copy)
@@ -1 +1 @@
-(define-constant +build-version+ "2.739")
+(define-constant +build-version+ "3.0.0rc1")
Index: runtime.c
===================================================================
--- runtime.c	(revision 7227)
+++ runtime.c	(working copy)
@@ -4192,6 +4192,9 @@
     return C_SCHEME_FALSE;
 #endif
 
+  case C_fix(41):
+    return C_fix(C_MAJOR_VERSION);
+
   default: return C_SCHEME_UNDEFINED;
   }
 }
Index: README
===================================================================
--- README	(revision 7227)
+++ README	(working copy)
@@ -2,7 +2,7 @@
   README file for the CHICKEN compiler
   (c)2000-2007 Felix L. Winkelmann
 
-  version 2.739
+  version 3.0.0rc1
 
 
  1. Introduction:
